libs/azure-storage/langchain_azure_storage/document_loaders.py

```diff
@@ -0,0 +1,112 @@
+    if isinstance(
+        credential,
+        (
+            azure.core.credentials.TokenCredential,
+            azure.core.credentials.AzureSasCredential,
+        ),
+    ):
+        return credential
+    raise TypeError("Invalid credential type provided.")
+
+
+class AzureBlobStorageLoader(BaseLoader):
+    """Document loader for loading LangChain Document objects from a container."""
+
+    def __init__(
+        self,
+        account_url: str,
+        container_name: str,
+        blob_names: Optional[Union[str, Iterable[str]]] = None,
+        prefix: str = "",
+        credential: Optional[
+            Union[
+                azure.core.credentials.AzureSasCredential,
+                azure.core.credentials.TokenCredential,
+            ]
+        ] = None,
+        loader_factory: Optional[Callable[[str], BaseLoader]] = None,
+    ):
+        """Initialize AzureBlobStorageLoader.
+
+        Parameters:
+        - account_url: str
+            URL to the Azure Storage account, e.g. "https://<account_name>.blob.core.windows.net"
+        - container_name: str
+            Name of the container to retrieve blobs from in the storage account
+        - blob_names: Optional[Union[str, Iterable[str]]]
+            List of blob names to load. If None, no blobs will be loaded
+        - prefix: str
+            Prefix to filter blobs in the container
+        - credential: Optional[Union[AzureSasCredential, TokenCredential]]
+            Credential to authenticate with the Azure Storage account.
+            If None, DefaultAzureCredential will be used
+        - loader_factory: Optional[Callable[[str], BaseLoader]]
+            A callable that returns a custom document loader to use for parsing blobs
+            downloaded. Ex: PyPDFLoader, CSVLoader, etc.
+        """
+        self._account_url = account_url
+        self._container_name = container_name
+        self._blob_names = [blob_names] if isinstance(blob_names, str) else blob_names
+        self._prefix = prefix
+        self._credential = credential
+        self._loader_factory = loader_factory
+
+    def lazy_load(self) -> Iterator[Document]:
+        """Lazily load documents from the specified blobs in the container."""
+        self._credential = _get_sdk_credential(self._credential)
+        if isinstance(
+            self._credential, azure.core.credentials_async.AsyncTokenCredential
+        ):
+            raise ValueError("Async credential provided to sync method.")
+        container_url = f"{self._account_url}/{self._container_name}"
+        container_client = ContainerClient.from_container_url(
+            container_url,
+            **_get_client_kwargs(self._credential),
+        )
+        blobs_with_prefix = list(
+            container_client.list_blob_names(name_starts_with=self._prefix)
+        )
+        if self._blob_names:
+            for blob_name in self._blob_names:
+                if blob_name in blobs_with_prefix:
```

Comment:
It would be interesting to encapsulate this logic into a helper method `_yield_blob_names()` that encapsulates both yielding from a listing blobs in the container and from the iterator. Mainly this will help remove some of the amount of nesting that is currently needed in the main method.